Prevent suspension for extension background pages

salinas-miguel opened the pull request:
#### d08338ef7145c7a3444b67ebdc3c4b72d63cab56
<pre>
Prevent suspension for extension background pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=247707">https://bugs.webkit.org/show_bug.cgi?id=247707</a>
rdar://problem/102167829

Reviewed by Chris Dumez.

Web content processes for web extension background pages are currently
being fully suspended when the WebKit:Suspended assertion is disabled.
We should disable full suspension when
_setPageVisibilityBasedProcessSuppressionEnabled is set to false as in
web extension background pages.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldTakeSuspendedAssertion):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/257628@main">https://commits.webkit.org/257628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a9306b18890f6934aedd9cb47c224f97657a3e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99372 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8571 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108756 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85889 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91862 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106683 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105130 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6909 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90452 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33884 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21799 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76765 "Found 4 new API test failures: /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/console-api, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/js-exception, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/network-error, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2439 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23316 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2345 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45712 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5249 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8505 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42791 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -631,13 +631,21 @@ void WebProcessProxy::addExistingWebPage(WebPageProxy& webPage, BeginsUsingDataS
     globalPageMap().set(webPage.identifier(), &webPage);
 
+    bool shouldTakeSuspendedAssertion = false;
+    for (auto& page : m_pageMap.values()) {
```
The code duplication is very unfortunate. We should factor this out to a separate function to avoid it.
Maybe a `shouldTakeSuspendedAssertion()` function on WebProcessProxy.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -648,6 +648,19 @@ Ref<WebPageProxy> WebProcessProxy::createWebPage(PageClient& pageClient, Ref<API
 }
 
+#if USE(RUNNINGBOARD)
+bool WebProcessProxy::shouldTakeSuspendedAssertion()
```
This should be marked as const.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -648,6 +648,19 @@ Ref<WebPageProxy> WebProcessProxy::createWebPage(PageClient& pageClient, Ref<API
     return webPage;
 }
 
+#if USE(RUNNINGBOARD)
```
May be nicer to guard the implementation with `#if USE(RUNNINGBOARD)` rather than the whole declaration, and have the function return false if we're not using running board. This would reduce ifdef'ing at call sites.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -581,6 +581,10 @@ class WebProcessProxy : public AuxiliaryProcessProxy, private ProcessThrottlerCl
 #endif
 
+#if USE(RUNNINGBOARD)
+    bool shouldTakeSuspendedAssertion();
```
const.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -581,6 +581,10 @@ class WebProcessProxy : public AuxiliaryProcessProxy, private ProcessThrottlerCl
     bool messageSourceIsValidWebContentProcess();
 #endif
 
+#if USE(RUNNINGBOARD)
```
Would be nice to avoid the `#if` here.

cdumez reviewed (approved):
LGTM with nits.

salinas-miguel committed:
Prevent suspension for extension background pages
https://bugs.webkit.org/show_bug.cgi?id=247707
rdar://problem/102167829

Reviewed by Chris Dumez.

Web content processes for web extension background pages are currently
being fully suspended when the WebKit:Suspended assertion is disabled.
We should disable full suspension when
_setPageVisibilityBasedProcessSuppressionEnabled is set to false as in
web extension background pages.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldTakeSuspendedAssertion):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: https://commits.webkit.org/257628@main
